File: src-tauri/src/cli/detector/makefile.rs
use crate::cli::detector::ProjectTemplate;
use std::fs;
use std::path::Path;

/// Detect Makefile targets
pub fn detect_makefile(directory: &Path) -> Option<Vec<ProjectTemplate>> {
    let makefile_names = ["Makefile", "makefile", "GNUmakefile"];
    let mut makefile_path = None;

    for name in &makefile_names {
        let path = directory.join(name);
        if path.exists() {
            makefile_path = Some(path);
            break;
        }
    }

    let makefile_path = makefile_path?;
    let content = fs::read_to_string(&makefile_path).ok()?;

    let mut projects = Vec::new();

    for line in content.lines() {
        // Detect common targets
        if let Some(target) = line.split(':').next() {
            let target = target.trim();

            // Skip special targets and variables
            if target.is_empty()
                || target.starts_with('.')
                || target.starts_with('#')
                || target.contains('=')
                || target.contains('$')
            {
                continue;
            }

            // Common targets to highlight
            let (_priority, description) = match target {
                "dev" | "develop" => (1, "Development server"),
                "start" | "run" | "serve" => (2, "Start application"),
                "build" | "compile" => (3, "Build project"),
                "test" | "tests" => (4, "Run tests"),
                "lint" | "linter" => (5, "Run linter"),
                "clean" => (100, "Clean build artifacts"),
                "install" | "deps" | "dependencies" => (6, "Install dependencies"),
                "watch" => (7, "Watch for changes"),
                "deploy" => (8, "Deploy application"),
                "fmt" | "format" => (9, "Format code"),
                "check" | "verify" => (10, "Verify/Check code"),
                _ => (50, "Makefile target"),
            };

            // Skip if already added
            if projects
                .iter()
                .any(|p: &ProjectTemplate| p.name == format!("make: {}", target))
            {
                continue;
            }

            projects.push(ProjectTemplate {
                name: format!("make: {}", target),
                command: format!("make {}", target),
                description: description.to_string(),
            });
        }
    }

    // Sort by priority
    projects.sort_by_key(|p| match p.name.as_str() {
        n if n.contains("dev") => 1,
        n if n.contains("start") || n.contains("run") || n.contains("serve") => 2,
        n if n.contains("build") => 3,
        n if n.contains("test") => 4,
        n if n.contains("lint") => 5,
        n if n.contains("install") || n.contains("deps") => 6,
        n if n.contains("watch") => 7,
        n if n.contains("deploy") => 8,
        n if n.contains("fmt") || n.contains("format") => 9,
        n if n.contains("check") || n.contains("verify") => 10,
        _ => 50,
    });

    if projects.is_empty() {
        None
    } else {
        Some(projects)
    }
}

File: src-electron/cli/detector/makefile.ts
/**
 * Makefile detector
 */

import * as fs from 'fs';
import * as path from 'path';
import type { ProjectTemplate } from '../types';

const MAKEFILE_NAMES = ['Makefile', 'makefile', 'GNUmakefile'];

export function detectMakefile(directory: string): ProjectTemplate[] | null {
  let makefilePath: string | null = null;

  for (const name of MAKEFILE_NAMES) {
    const fullPath = path.join(directory, name);
    if (fs.existsSync(fullPath)) {
      makefilePath = fullPath;
      break;
    }
  }

  if (!makefilePath) {
    return null;
  }

  try {
    const content = fs.readFileSync(makefilePath, 'utf-8');
    const projects: ProjectTemplate[] = [];
    const seenTargets = new Set<string>();

    for (const line of content.split('\n')) {
      // Parse target: dependencies
      const colonIndex = line.indexOf(':');
      if (colonIndex === -1) continue;

      const target = line.substring(0, colonIndex).trim();

      // Skip special targets and variables
      if (!target || target.startsWith('.') || target.startsWith('#')) continue;
      if (target.includes('=') || target.includes('$')) continue;
      if (seenTargets.has(target)) continue;

      seenTargets.add(target);

      const { description, priority } = getMakefileTargetInfo(target);

      projects.push({
        name: `make: ${target}`,
        command: `make ${target}`,
        description,
        priority
      });
    }

    // Sort by priority
    projects.sort((a, b) => (a.priority || 50) - (b.priority || 50));

    return projects.length > 0 ? projects : null;
  } catch {
    return null;
  }
}

function getMakefileTargetInfo(target: string): { description: string; priority: number } {
  const targetMap: Record<string, { description: string; priority: number }> = {
    'dev': { description: 'Development server', priority: 1 },
    'develop': { description: 'Development server', priority: 1 },
    'start': { description: 'Start application', priority: 2 },
    'run': { description: 'Start application', priority: 2 },
    'serve': { description: 'Start application', priority: 2 },
    'build': { description: 'Build project', priority: 3 },
    'compile': { description: 'Build project', priority: 3 },
    'test': { description: 'Run tests', priority: 4 },
    'tests': { description: 'Run tests', priority: 4 },
    'lint': { description: 'Run linter', priority: 5 },
    'linter': { description: 'Run linter', priority: 5 },
    'install': { description: 'Install dependencies', priority: 6 },
    'deps': { description: 'Install dependencies', priority: 6 },
    'dependencies': { description: 'Install dependencies', priority: 6 },
    'watch': { description: 'Watch for changes', priority: 7 },
    'deploy': { description: 'Deploy application', priority: 8 },
    'fmt': { description: 'Format code', priority: 9 },
    'format': { description: 'Format code', priority: 9 },
    'check': { description: 'Verify/Check code', priority: 10 },
    'verify': { description: 'Verify/Check code', priority: 10 },
    'clean': { description: 'Clean build artifacts', priority: 100 }
  };

  return targetMap[target] || { description: 'Makefile target', priority: 50 };
}
